### PR TITLE
Testing book functionality

### PR DESCRIPTION
**Pull request by ShastkivRuslan**

### CODE REVIEW

File: src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        );
+
+        MvcResult result = mockMvc.perform(
+                        MockMvcRequestBuilders.get(BOOK_URL)
+                                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andReturn();
+
+        String expectedString = objectMapper.writeValueAsString(expectedPage);
+        String actualString = result.getResponse().getContentAsString();
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java]: can you compare objects?

File: src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        assertEquals(expected, actual);
+    }
+
+    @WithMockUser(username = "user")
+    @Test
+    @DisplayName("""
+            Should throw EntityNotFoundException for an invalid book ID
+            """)
+    public void getBokById_BookByInvalidId_ThrowException() throws Exception {
+        System.out.println(this);
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java]:
```suggestion
```

File: src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+        MvcResult result = mockMvc.perform(
+                        MockMvcRequestBuilders.get(SEARCH_BOOK_URL)
+                                .param("authors", CUSTOM_BOOK_AUTHOR.formatted(FIRST_BOOK_ID))
+                                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andReturn();
+
+        String expectedString = objectMapper.writeValueAsString(expectedPage);
+        String actualString = result.getResponse().getContentAsString();
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java]: the same

File: src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+        assertEquals(expectedString, actualString);
+
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java]:
```suggestion
```

File: src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+                        MockMvcRequestBuilders.get(SEARCH_BOOK_URL)
+                                .param("titles", NON_EXISTING_TITLE)
+                                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andReturn();
+        Page<BookDto> expectedPage = new PageImpl<>(List.of(), PAGEABLE, EMPTY_PAGE);
+        String expectedString = objectMapper.writeValueAsString(expectedPage);
+        String actualString = result.getResponse().getContentAsString();
+
+        assertEquals(expectedString, actualString);
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/controller/BookControllerTest.java]: Let’s  follow the Given-When-Then test structure

File: src/test/java/ruslan/shastkiv/bookstore/controller/CategoryControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+},
+        executionPhase = Sql.ExecutionPhase.BEFORE_TEST_CLASS)
+@Sql(scripts = {
+        "classpath:scripts/book/remove_relations_between_books_and_categories_from_db.sql",
+        "classpath:scripts/category/remove_categories_from_db.sql",
+        "classpath:scripts/book/remove_test_books_from_db.sql"
+},
+        executionPhase = Sql.ExecutionPhase.AFTER_TEST_CLASS)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+public class CategoryControllerTest {
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/controller/CategoryControllerTest.java:54]: Let’s apply the same comments ;)

File: src/test/java/ruslan/shastkiv/bookstore/controller/CategoryControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+                                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andReturn();
+
+        CategoryDto expected = createCategoryDtoById(FIRST_CATEGORY_ID);
+        CategoryDto actual = objectMapper.readValue(
+                result.getResponse().getContentAsString(), CategoryDto.class);
+
+        assertEquals(expected, actual);
+
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/controller/CategoryControllerTest.java:135]:
```suggestion
```

File: src/test/java/ruslan/shastkiv/bookstore/repository/BookRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            Should fetch books with initialized categories
+            """)
+    void findAllWithCategories_getInitializedCategory_DoesNotThrowLazyInitializationException() {
+        List<Book> allWithCategories = bookRepository.findAllWithCategories(PAGEABLE).getContent();
+
+        for (Book book : allWithCategories) {
+            assertDoesNotThrow(() -> {
+                book.getCategories().size();
+            }, "LazyInitializationException was thrown for book: " + book.getTitle());
+        }
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/repository/BookRepositoryTest.java]: you can refactor it 
```
assertAll(
    allWithCategories.stream()
        .map(book -> () -> assertDoesNotThrow(
            () -> book.getCategories().size(),
            "LazyInitializationException was thrown for book: " + book.getTitle()
        ))
);
```

File: src/test/java/ruslan/shastkiv/bookstore/service/BookServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+import ruslan.shastkiv.bookstore.service.book.BookServiceImpl;
+
+@ExtendWith(MockitoExtension.class)
+public class BookServiceTest {
+    @InjectMocks
+    private BookServiceImpl bookService;
+    @Mock
+    private BookRepository bookRepository;
+    @Mock
+    private BookMapper bookMapper;
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/service/BookServiceTest.java:58]: Advanced (optional): you can use Spy and simplify your test cases

File: src/test/java/ruslan/shastkiv/bookstore/service/BookServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+            Throw EntityNotFoundException when book ID does not exist
+            """)
+    void findBookById_InvalidId_ThrowException() {
+        when(bookRepository.findById(INVALID_BOOK_ID)).thenReturn(Optional.empty());
+
+        assertThrows(EntityNotFoundException.class, ()
+                -> bookService.findBookById(INVALID_BOOK_ID));
+    }
+}
+
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/service/BookServiceTest.java:254]:
```suggestion
```

File: src/test/java/ruslan/shastkiv/bookstore/service/CategoryServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+        when(categoryMapper.toDto(firstCategory)).thenReturn(firstCategoryDto);
+        when(categoryMapper.toDto(secondCategory)).thenReturn(secondCategoryDto);
+
+        Page<CategoryDto> actual = categoryService.getAllCategories(PAGEABLE);
+
+        assertEquals(PAGE_SIZE_2, actual.getTotalElements());
+        assertEquals(PAGEABLE.getPageSize(), actual.getSize());
+        assertEquals(List.of(firstCategoryDto, secondCategoryDto), actual.getContent());
+
+        // Verify interactions
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/service/CategoryServiceTest.java]:
```suggestion
```

File: src/test/java/ruslan/shastkiv/bookstore/service/CategoryServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+    @Test
+    @DisplayName("Should delete a category by its ID")
+    public void deleteCategory_GivenValidId_DeleteCategory() {
+        categoryService.deleteCategory(FIRST_CATEGORY_ID);
+
+        verify(categoryRepository).deleteById(FIRST_CATEGORY_ID);
+        verifyNoMoreInteractions(categoryRepository);
+    }
+
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/service/CategoryServiceTest.java:170]:
```suggestion
```

File: src/test/java/ruslan/shastkiv/bookstore/utils/BookTestUtils.java [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.stream.Collectors;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import ruslan.shastkiv.bookstore.dto.book.BookDto;
+import ruslan.shastkiv.bookstore.dto.book.BookDtoWithoutCategoryIds;
+import ruslan.shastkiv.bookstore.dto.book.BookSearchParametersDto;
+import ruslan.shastkiv.bookstore.dto.book.CreateBookRequestDto;
+import ruslan.shastkiv.bookstore.model.Book;
+
+public class BookTestUtils {
```
Comment (liliia-ponomarenko) [inline src/test/java/ruslan/shastkiv/bookstore/utils/BookTestUtils.java:20]: 🔥

File: src/test/resources/scripts/book/insert_relations_between_book_and_categories.sql [hunk 14]
```diff
@@ -0,0 +1,4 @@
+INSERT INTO books_categories (book_id, category_id) VALUES
+    (1, 1),
+    (2, 2),
+    (3, 2);
```
Comment (liliia-ponomarenko) [inline src/test/resources/scripts/book/insert_relations_between_book_and_categories.sql]: add one empty line to the end of the file (check other files)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  improve your solution ;)

**Comment by ShastkivRuslan** [comment]
Object comparison: Replaced string comparison of expectedString and actualString with proper object comparison in the test cases. Now the objects are compared directly, ensuring correctness and better readability.

Test structure improvement: Refactored tests to follow the Given-When-Then structure for better clarity and consistency in test case formatting.

Code refactoring: Simplified assertions and refactored repetitive code:

In BookRepositoryTest, replaced individual assertions with a more compact assertAll block to check for LazyInitializationException.
Simplified mock verifications and removed redundant interactions in tests.

Small formatting fixes: Added an empty line at the end of insert_relations_between_book_and_categories.sql as per standard project formatting practices.

**Review by sarakhmen — APPROVED** [review]
Great job!

**Pull request merged**